hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java

```diff
@@ -568,15 +563,21 @@ protected boolean containsKeyInternal(Data key) {
     public void waitUntilLoaded() {
         try {
             int mapNamePartition = partitionService.getPartitionId(name);
-            Operation op = new PartitionCheckIfLoadedOperation(name, false, true);
-            Future loadingFuture = operationService.invokeOnPartition(SERVICE_NAME, op, mapNamePartition);
-            // wait for keys to be loaded - it's insignificant since it doesn't trigger the keys loading
-            // it's just waiting for them to be loaded. Timeout failure doesn't mean anything negative here.
-            // This call just introduces some ordering of requests.
-            FutureUtil.waitWithDeadline(singleton(loadingFuture), CHECK_IF_LOADED_TIMEOUT_SECONDS, SECONDS,
-                    logAllExceptions(WARNING));
-
-            OperationFactory opFactory = new PartitionCheckIfLoadedOperationFactory(name);
+            // first we have to check if key-load finished - otherwise the loading on other partitions might not have started.
+            // In this case we can't invoke IsPartitionLoadedOperation -> they will return "true", but it won't be correct.
+
+            int sleepSeconds = 1;
+            while (true) {
+                Operation op = new IsKeyLoadFinishedOperation(name);
+                Future<Boolean> loadingFuture = operationService.invokeOnPartition(SERVICE_NAME, op, mapNamePartition);
+                if (loadingFuture.get()) {
+                    break;
+                }
+                // sleep with some back-off
+                TimeUnit.SECONDS.sleep(sleepSeconds < MAXIMAL_WAIT_LOAD_SLEEP_SECS ? ++sleepSeconds : sleepSeconds);
```

Comment:
this could be seen as a performance regression.

 when the first check is unsuccessful then it will wait at least 1 more second. even when the actual map-loading is shorter. I think this was the original implementation in the `PartitionCheckIfLoadedOperation`, but it was changed to prevent this extra waiting. let me find the issue. see https://github.com/hazelcast/hazelcast/issues/5349